### PR TITLE
Allow ending polls

### PR DESCRIPTION
**Pull request by andybalaam**
This change allows us to end a poll and see the results highlighted and with a little trophy symbol:

end-poll

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

Preview: https://61b0c70c1a94ab2550cb7837--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

File: res/css/views/messages/_MPollBody.scss [hunk 1]
```diff
@@ -49,11 +49,11 @@ limitations under the License.
         max-width: 550px;
         background-color: $background;
 
-        .mx_StyledRadioButton {
+        .mx_StyledRadioButton,.mx_MPollBody_endedOption {
```
Comment (t3chguy) [inline res/css/views/messages/_MPollBody.scss:52]:
```suggestion
        .mx_StyledRadioButton, .mx_MPollBody_endedOption {
```

File: res/css/views/messages/_MPollBody.scss [hunk 2]
```diff
@@ -49,11 +49,11 @@ limitations under the License.
...
             margin-bottom: 8px;
         }
 
-        .mx_StyledRadioButton_content {
+        .mx_StyledRadioButton_content,.mx_MPollBody_endedOption {
```
Comment (t3chguy) [inline res/css/views/messages/_MPollBody.scss:56]:
```suggestion
        .mx_StyledRadioButton_content, .mx_MPollBody_endedOption {
```

File: res/css/views/messages/_MPollBody.scss [hunk 3]
```diff
@@ -102,20 +102,37 @@ limitations under the License.
...
-        border-color: $accent;
-        background-color: $accent;
-        background-image: url('$(res)/img/element-icons/check-white.svg');
-        background-size: 12px;
-        background-repeat: no-repeat;
-        background-position: center;
-
-        div {
-            visibility: hidden;
+    .mx_StyledRadioButton_checked,.mx_MPollBody_endedOptionWinner {
```
Comment (t3chguy) [inline res/css/views/messages/_MPollBody.scss:105]:
```suggestion
    .mx_StyledRadioButton_checked, .mx_MPollBody_endedOptionWinner {
```

File: src/components/views/context_menus/MessageContextMenu.tsx [hunk 4]
```diff
@@ -68,6 +72,11 @@ interface IProps extends IPosition {
     onFinished(): void;
     /* if the menu is inside a dialog, we sometimes need to close that dialog after click (forwarding) */
     onCloseDialog?(): void;
+    getRelationsForEvent?: (
+            eventId: string,
+            relationType: string,
+            eventType: string
+        ) => Relations;
```
Comment (t3chguy) [inline src/components/views/context_menus/MessageContextMenu.tsx:79]:
```suggestion
    getRelationsForEvent?: (
        eventId: string,
        relationType: string,
        eventType: string
    ) => Relations;
```

File: src/components/views/context_menus/MessageContextMenu.tsx [hunk 5]
```diff
@@ -123,6 +132,14 @@ export default class MessageContextMenu extends React.Component<IProps, IState>
         return content.pinned && Array.isArray(content.pinned) && content.pinned.includes(this.props.mxEvent.getId());
     }
 
+    private canEndPoll(mxEvent: MatrixEvent) {
```
Comment (t3chguy) [inline src/components/views/context_menus/MessageContextMenu.tsx:135]:
```suggestion
    private canEndPoll(mxEvent: MatrixEvent): boolean {
```

File: src/components/views/dialogs/EndPollDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+interface IProps extends IDialogProps {
+    matrixClient: MatrixClient;
+    event: MatrixEvent;
+    onFinished: (success: boolean) => void;
+    getRelationsForEvent?: (
+            eventId: string,
+            relationType: string,
+            eventType: string
+        ) => Relations;
```
Comment (t3chguy) [inline src/components/views/dialogs/EndPollDialog.tsx:38]:
```suggestion
        eventId: string,
        relationType: string,
        eventType: string
    ) => Relations;
```

File: src/components/views/dialogs/EndPollDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+import { MatrixClient } from "matrix-js-sdk/src/client";
+
+import { _t } from "../../../languageHandler";
+import { IDialogProps } from "./IDialogProps";
+import QuestionDialog from "./QuestionDialog";
+import { IPollEndContent, POLL_END_EVENT_TYPE, TEXT_NODE_TYPE } from "../../../polls/consts";
+import { findTopAnswer } from "../messages/MPollBody";
+import Modal from "../../../Modal";
+import ErrorDialog from "./ErrorDialog";
+import { Relations } from "matrix-js-sdk/src/models/relations";
```
Comment (t3chguy) [inline src/components/views/dialogs/EndPollDialog.tsx:28]: could this go with the other js-sdk imports?

File: src/components/views/messages/MessageActionBar.tsx [hunk 8]
```diff
@@ -51,10 +51,22 @@ interface IOptionsButtonProps {
     getReplyChain: () => ReplyChain;
     permalinkCreator: RoomPermalinkCreator;
     onFocusChange: (menuDisplayed: boolean) => void;
+    getRelationsForEvent?: (
+            eventId: string,
+            relationType: string,
+            eventType: string
+        ) => Relations;
```
Comment (t3chguy) [inline src/components/views/messages/MessageActionBar.tsx:58]:
```suggestion
        eventId: string,
        relationType: string,
        eventType: string
    ) => Relations;
```

File: src/components/views/messages/MessageActionBar.tsx [hunk 9]
```diff
@@ -51,10 +51,22 @@ interface IOptionsButtonProps {
...
-    ({ mxEvent, getTile, getReplyChain, permalinkCreator, onFocusChange }) => {
+    ({
+        mxEvent,
+        getTile,
+        getReplyChain,
+        permalinkCreator,
+        onFocusChange,
+        getRelationsForEvent,
+    }) => {
         const [menuDisplayed, button, openMenu, closeMenu] = useContextMenu();
```
Comment (t3chguy) [inline src/components/views/messages/MessageActionBar.tsx:70]: this indentation feels wrong

File: src/components/views/messages/MessageActionBar.tsx [hunk 10]
```diff
@@ -138,6 +151,11 @@ interface IMessageActionBarProps {
     onFocusChange?: (menuDisplayed: boolean) => void;
     toggleThreadExpanded: () => void;
     isQuoteExpanded?: boolean;
+    getRelationsForEvent?: (
+            eventId: string,
+            relationType: string,
+            eventType: string
+        ) => Relations;
```
Comment (t3chguy) [inline src/components/views/messages/MessageActionBar.tsx:158]:
```suggestion
        eventId: string,
        relationType: string,
        eventType: string
    ) => Relations;
```

File: src/components/views/messages/MPollBody.tsx [hunk 11]
```diff
@@ -283,13 +439,58 @@ export default class MPollBody extends React.Component<IBodyProps, IState> {
...
+        <div className="mx_MPollBody_optionDescription">
+            <div className="mx_MPollBody_optionText">
+                { answer[TEXT_NODE_TYPE.name] }
+            </div>
+            <div className="mx_MPollBody_optionVoteCount">
+                { votesText }
+            </div>
+        </div>
+    </StyledRadioButton>;
+}
```
Comment (t3chguy) [inline src/components/views/messages/MPollBody.tsx:485]: These should be refactored into Functional components to be more akin to React patterns.

PascalCased name, props should be an object, and change callee to instantiate using JSX.

File: src/components/views/messages/MPollBody.tsx [hunk 12]
```diff
@@ -299,16 +500,67 @@ function userResponseFromPollResponseEvent(event: MatrixEvent): UserVote {
...
-        return pollRelations.getRelations()
+export function allVotes(
+    pollEvent: MatrixEvent,
+    matrixClient: MatrixClient,
+    voteRelations: Relations,
+    endRelations: Relations,
+): Array<UserVote> {
+    const endTs = pollEndTs(pollEvent, matrixClient, endRelations);
+
+    function isOnOrBeforeEnd(responseEvent: MatrixEvent) {
```
Comment (t3chguy) [inline src/components/views/messages/MPollBody.tsx:511]:
```suggestion
    function isOnOrBeforeEnd(responseEvent: MatrixEvent): boolean {
```

File: src/components/views/messages/MPollBody.tsx [hunk 13]
```diff
@@ -299,16 +500,67 @@ function userResponseFromPollResponseEvent(event: MatrixEvent): UserVote {
...
 }
 
+/**
+ * Returns the earliest timestamp from the supplied list of end_poll events.
+ */
+export function pollEndTs(
+    pollEvent: MatrixEvent,
+    matrixClient: MatrixClient,
+    endRelations: Relations,
+) {
```
Comment (t3chguy) [inline src/components/views/messages/MPollBody.tsx:537]: return type?

**Review by t3chguy — CHANGES_REQUESTED** [review]

File: src/components/views/messages/MessageActionBar.tsx (hunk 9 above)
Comment (andybalaam) [reply src/components/views/messages/MessageActionBar.tsx:70]: Is this better?

```
const OptionsButton: React.FC<IOptionsButtonProps> =
    ({
        mxEvent, getTile, getReplyChain, permalinkCreator, onFocusChange,
        getRelationsForEvent,
    }) => {
        const [menuDisplayed, button, openMenu, closeMenu] = useContextMenu();
```

Or what would you prefer?

File: src/components/views/messages/MPollBody.tsx (hunk 11 above)
Comment (andybalaam) [reply src/components/views/messages/MPollBody.tsx:485]: Thank you for the detailed review, and sorry there was a lot to fix!

File: src/components/views/messages/MessageActionBar.tsx (hunk 9 above)
Comment (t3chguy) [reply src/components/views/messages/MessageActionBar.tsx:70]: IMO the `const [menuDisplayed, button, openMenu, closeMenu] = useContextMenu();` should only be one-level indented from the `const OptionsButton: React.FC<IOptionsButtonProps>`

We normally do 

```
const OptionsButton: React.FC<IOptionsButtonProps> = ({
    prop,
    ...
}) => {
    code
    ...
};

File: src/components/views/messages/MPollBody.tsx (hunk 13 above)
Comment (andybalaam) [reply src/components/views/messages/MPollBody.tsx]: Added in ee129db4fb

File: src/components/views/messages/MessageActionBar.tsx (hunk 9 above)
Comment (andybalaam) [reply src/components/views/messages/MessageActionBar.tsx:70]: Ah yeah I prefer that too.

File: src/components/views/messages/MPollBody.tsx (hunk 11 above)
Comment (andybalaam) [reply src/components/views/messages/MPollBody.tsx]: Had a go at this in f90b28d398

File: src/components/views/messages/MessageActionBar.tsx (hunk 9 above)
Comment (andybalaam) [reply src/components/views/messages/MessageActionBar.tsx]: Done in 9e8786cb6e.

This should be ready for re-review, thank you.

**Review by gaelledel — APPROVED** [review]
Good stuff!
I wonder if we could tweak the radio button colour when unchecked to Dark/Quinary? https://www.figma.com/file/tepA3ELWBE5v4O2WB8PqWn/Polls?node-id=496%3A54285

And for the votes text to be right aligned with the bar with equal number of padding that you have on the left side. If it is 8px currently, can we do 12px on each side instead?
Screenshot 2021-12-08 at 12 31 07
Screenshot 2021-12-08 at 12 31 23

**Comment by andybalaam** [comment]
> I wonder if we could tweak the radio button colour when unchecked to Dark/Quinary? figma.com/file/tepA3ELWBE5v4O2WB8PqWn/Polls?node-id=496%3A54285

Should it be Quinary in the Settings dialog too?  They share a colour at the moment:
image

**Comment by andybalaam** [comment]
> And for the votes text to be right aligned with the bar with equal number of padding that you have on the left side. If it is 8px currently, can we do 12px on each side instead?

That's weird - it looks right for me:
image

Maybe it's a cross-browser issue.  I'll look into it.

**Comment by andybalaam** [comment]
> Maybe it's a cross-browser issue. I'll look into it.

Ah no, my bad - it goes wrong when the poll is ended.  Working on it.

**Comment by andybalaam** [comment]
OK, @gaelledel I think I've completed the changes you asked for.
image
image
image
image

File: src/components/views/dialogs/EndPollDialog.tsx (hunk 7 above)
Comment (andybalaam) [reply src/components/views/dialogs/EndPollDialog.tsx]: Done in bfd594ab0d

**Comment by gaelledel** [comment]
> > I wonder if we could tweak the radio button colour when unchecked to Dark/Quinary? figma.com/file/tepA3ELWBE5v4O2WB8PqWn/Polls?node-id=496%3A54285
> 
> Should it be Quinary in the Settings dialog too? They share a colour at the moment: image
Is it the same colour for Light and Dark ? E.g: Quartenary/Light - Quarternary/Dark?

**Comment by andybalaam** [comment]
> > Is it the same colour for Light and Dark ? E.g: Quartenary/Light - Quarternary/Dark?

Yes, in both themes it is quaternary.  My latest change makes it quinary in both.  Is that right?

**Comment by andybalaam** [comment]
> Yes, in both themes it is quaternary. My latest change makes it quinary in both. Is that right?

We agreed to go back to quaternary for both themes.  Done in f697a17718

File: src/components/views/messages/MPollBody.tsx [hunk 14]
```diff
@@ -439,46 +441,50 @@ export default class MPollBody extends React.Component<IBodyProps, IState> {
...
+    answer: IPollAnswer;
+    checked: boolean;
+    votesText: string;
+}
+
+function EndedPollOption(props: IEndedPollOptionProps) {
     const classNames = `mx_MPollBody_endedOption${
-        checked ? " mx_MPollBody_endedOptionWinner": ""
+        props.checked ? " mx_MPollBody_endedOptionWinner": ""
     }`;
```
Comment (t3chguy) [inline src/components/views/messages/MPollBody.tsx]: can we use the classNames utility package here plase?

**Review by t3chguy — APPROVED** [review]
LGTM otherwise

File: src/components/views/messages/MPollBody.tsx (hunk 14 above)
Comment (andybalaam) [reply src/components/views/messages/MPollBody.tsx]: Ah, I didn't know about that - thanks, done in 79e1d45d925270e2f65e2f1f030c7f502a938eaf

**Pull request merged**